### ISSUE TITLE
skmine.metrics

### ISSUE THREAD

**Issue by remiadon** [scikit-mine/scikit-mine#77]
#### Describe the workflow you want to enable
scikit-learn provides a `sklearn.metrics` module which help to "quantify the quality of predictions"

In our framework we might want to quantify the quality of the patterns we find

#### Describe your proposed solution
Instantiate a `skmine.metrics` module with a few basic metrics, for post-hoc ranking of our beloved patterns
As we return pandas containers in scikit-mine, the metric API should encourage nifty syntaxes, like

```python
>> from skmine.some_module import Miner
>> from skmine.metric import some_metric
>> miner = Miner()
>> D = fetch_some_dataset()
>> patterns = miner.fit_discover(D)  # or miner.fit(D).codetable if miner is MDL based
>> list(patterns.columns)
>> ['itemset', 'support']
>> indicator = some_metric(patterns.support)  # returns a vector
>> patterns.loc[:, 'indicator'] = indicator
>> patterns.nlargest(20, 'indicator')
```

@PeggyCellier @alexandre-termier have you got a reference in mind that lists the standard metrics for pattern mining ?

**Comment by remiadon** [comment]
@PeggyCellier  metionned [this survey](https://dl.acm.org/doi/pdf/10.1145/1132960.1132963) for a good overview of metrics

**Comment by lgalarra** [comment]
Hi Rémi,

Thanks, I think I knew that paper. Indeed, interesting measures depend on the type of the patterns. For example confidence is defined for rules, neither for itemsets nor for sequences. In that line it is the miner that should provide the interesting measure. I also agree that ranking and filter metrics should be a post-hoc process, so as measures of statistical significance. We could therefore envision something like this:

```python
>> from skmine.some_module import Miner
>> from skmine.some_module import confidence_intervals, p_value
>> D = fetch_data()
>> miner = Miner()
>> patterns = miner.fit_discover()
>> list(patterns.columns)
>> ['itemset', 'a_metric', 'a_cooler_metric']
>> supp, inf = confidence_intervals(patterns, metric='a_cooler_metric', data=D, significance=0.95)
>> p = p_value(patterns, metric='a_cooler_metric')
>> patterns.loc[:, 'upper_bound_cooler_metric'] = supp
>> patterns.loc[:, 'lower_bound_cooler_metric'] = inf
>> patterns.loc[:, 'p-value_cooler_metric'] = p 
```
That is, the module provides that functionality for the metrics where it is pertinent (as I said, confidence intervals and p_values make sense for certain metrics). 

Cheers,
Luis

**Comment by remiadon** [comment]
@lgalarra thanks, but I need some clarifications

In your example, how did you get `a_cooler_metric` ? Can we compute it from `a_metric` (`a_metric` being the support, for example) ?

If we manage to compute every metric from the support, this will become way more elegant in terms of API and modularity

**Comment by lgalarra** [comment]
In that snippet I am assuming that the miner already computes 'a_cooler_metric' and that it may or may not be calculable from 'a_metric'. For instance if 'a_cooler_metric' is the standard confidence in ARM, having the support ('a_metric') of A=>B alone is not enough, as you also need the support of B. In that particular example I think it would be more elegant and useful to return the confidence from the miner instead of having to normalize the support afterwards.
